safehttp/form_test.go

```diff
@@ -12,1135 +12,476 @@
 	tests := []struct {
-		req         *http.Request
+		name        string
+		values      []string
 		placeholder interface{}
-		want        interface{}
 	}{
 		{
-			req: func() *http.Request {
-				req := httptest.NewRequest("POST", "http://foo.com/", strings.NewReader("pizza=-8&pizza=9&pizza=-100"))
-				req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
-				return req
-			}(),
+			name:        "Int64",
+			values:      []string{"1", "abc", "1"},
 			placeholder: &[]int64{},
-			want:        &[]int64{-8, 9, -100},
 		},
 		{
-			req: func() *http.Request {
-				req := httptest.NewRequest("POST", "http://foo.com/", strings.NewReader("pizza=8&pizza=9&pizza=10"))
-				req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
-				return req
-			}(),
+			name:        "Uint64",
+			values:      []string{"1", "abc", "-1"},
 			placeholder: &[]uint64{},
-			want:        &[]uint64{8, 9, 10},
 		},
 		{
-			req: func() *http.Request {
-				req := httptest.NewRequest("POST", "http://foo.com/", strings.NewReader("pizza=margeritta&pizza=diavola&pizza=calzone"))
-				req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
-				return req
-			}(),
-			placeholder: &[]string{},
-			want:        &[]string{"margeritta", "diavola", "calzone"},
-		},
-		{
-			req: func() *http.Request {
-				req := httptest.NewRequest("POST", "http://foo.com/", strings.NewReader("pizza=1.3&pizza=8.9&pizza=-4.1"))
-				req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
-				return req
-			}(),
+			name:        "Float64",
+			values:      []string{"1.3", "abc", "-4.1"},
 			placeholder: &[]float64{},
-			want:        &[]float64{1.3, 8.9, -4.1},
 		},
 		{
-			req: func() *http.Request {
-				req := httptest.NewRequest("POST", "http://foo.com/", strings.NewReader("pizza=true&pizza=false&pizza=T"))
-				req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
-				return req
-			}(),
+			name:        "Bool",
+			values:      []string{"t", "abc", "TRUE"},
 			placeholder: &[]bool{},
-			want:        &[]bool{true, false, true},
 		},
 	}
 
-	for _, tc := range tests {
-		mux := safehttp.NewServeMux(testDispatcher{}, "foo.com")
-		mux.Handle("/", safehttp.MethodPost, safehttp.HandlerFunc(func(rw safehttp.ResponseWriter, ir *safehttp.IncomingRequest) safehttp.Result {
-			form, err := ir.PostForm()
-			if err != nil {
-				t.Fatalf(`ir.PostForm: got err %v, want nil`, err)
-			}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			f := Form{values: map[string][]string{"x": tt.values}}
 
-			form.Slice("pizza", tc.placeholder)
-			if err := form.Err(); err != nil {
-				t.Errorf(`form.Err: got err %v, want nil`, err)
-			}
-
-			if diff := cmp.Diff(tc.want, tc.placeholder); diff != "" {
-				t.Errorf("form.Slice: diff (-want +got): \n%v", diff)
-			}
+			f.Slice("x", tt.placeholder)
 
-			return safehttp.Result{}
-		}))
+			// TODO: add a check here that tt.placeholder is nil. I (grenfeldt@)
```

Comment:
Keep the placeholder as an interface and do a `cmp.Diff` with a nil interface (`var want interface{}`).

Reply:
So, this is very tricky. If i have something like this:
```go
var want interface{}
diff := cmp.Diff(want, tt.placeholder)
```
I get the following diff:
```
form_test.go:363: DIFF (-want +got)  interface{}(
            + 	&[]int64(nil),
              )
```
If I try to create a `&[]int(nil)` to compare with in the table then it complains that I "can't take the address of nil".
If I try to create a `[]int(nil)` to compare with then that works.
```go
diff := cmp.Diff(&tt.want, tt.placeholder)
```
But then I get the following diff which compares pointer addresses :/
```
form_test.go:363: DIFF (-want +got)   interface{}(
            - 	&⟪0xc000066768⟫[]int(nil),
            +  &⟪0xc00000f120⟫[]int(nil),
              )
```
And if instead to try dereference `tt.placeholder` then that is really annoying as well since you have to do a bunch of type assertions.
Maybe there is some `cmpopt` that let's me compare pointers without comparing their addresses. I'll look into that.
Does @empijei or @kele have any tips on this?

Reply:
Played around with it a bit and realised the solution I thought I had from some weeks ago doesn't work :(

Reply:
I'm not sure I get why you want a `nil` placeholder check. What's the rationale?

Reply:
`f.Slice` will set the slice pointer to `nil` if the data was invalid, I want to check that it does.